File: core/state_machine.py
"""状态机：执行状态转换，记录日志
State machine: execute state transitions, record logs.

转换表完全由工作流注册表提供，框架不内置任何业务状态
Transition table is entirely provided by the workflow registry; the framework has no built-in business states."""

from __future__ import annotations

from core.db import _TABLE_COLUMNS, atomic_transaction, merge_extra_json, now


class InvalidTransitionError(Exception):
    pass


def _resolve_transitions(task_id: str, conn) -> dict[str, list[tuple[str, str]]]:
    """解析任务对应的转换表。
    Resolve the transition table for the given task.

    从工作流注册表查询，查不到返回空字典。
    Queries the workflow registry; returns empty dict if not found."""
    row = conn.execute("SELECT workflow FROM tasks WHERE id = ?", (task_id,)).fetchone()
    if not row:
        return {}

    workflow_name = row["workflow"] if row["workflow"] else ""

    try:
        from core import registry

        transitions = registry.build_transitions(workflow_name)
        if transitions:
            return transitions
    except ImportError:
        pass
    except Exception as e:
        import logging

        logging.getLogger(__name__).warning("解析转换表失败（workflow=%s）：%s", workflow_name, e)

    return {}


def transition(
    task_id: str,
    trigger: str,
    note: str | None = None,
    extra_updates: dict | None = None,
    transitions: dict | None = None,
) -> tuple[str, str]:
    """执行状态转换（原子操作，完全手动事务管理）
    Execute state transition (atomic operation, fully manual transaction management).

    Args:
        task_id: 任务 ID / Task ID
        trigger: 触发器名称 / Trigger name
        note: 可选备注 / Optional note
        extra_updates: 额外要更新的字段 dict（如 rejection_counts）
                       Additional fields to update (e.g. rejection_counts)
        transitions: 可选，外部传入的转换表。不传时自动从任务的 workflow 查注册表。
                     Optional externally provided transition table.
                     Auto-resolved from workflow registry if not provided.

    Returns:
        (from_status, to_status)

    Raises:
        InvalidTransitionError: 非法状态转换 / Invalid state transition
    """
    with atomic_transaction() as conn:
        row = conn.execute("SELECT * FROM tasks WHERE id = ?", (task_id,)).fetchone()
        if not row:
            raise ValueError(f"任务不存在：{task_id}")

        from_status = row["status"]

        # 确定使用的转换表 / Determine transition table to use
        if transitions is None:
            transitions = _resolve_transitions(task_id, conn)

        # 查找合法目标状态 / Find valid target state
        allowed = transitions.get(from_status, [])
        to_status = None
        for t, dest in allowed:
            if t == trigger:
                to_status = dest
                break

        if to_status is None:
            raise InvalidTransitionError(
                f"非法转换：{from_status} --[{trigger}]--> ??? （允许的 trigger：{[t for t, _ in allowed]}）"
            )

        # 构建更新字段（透明区分列字段 vs extra JSON）
        # Build update fields (transparently distinguish column fields vs extra JSON)
        col_updates: dict = {"status": to_status, "updated_at": now()}
        # 仅在进入 running 状态时更新 started_at，保持"阶段开始时间"语义
        # Only update started_at when entering running state to preserve "phase start time" semantics
        is_running = False
        try:
            from core import registry

            running_map = registry.get_running_state_phase(row["workflow"] or "")
            is_running = to_status in running_map
        except Exception:
            # fallback：按命名模式判断 / Fallback: check by naming pattern
            is_running = "running" in to_status
        if is_running:
            col_updates["started_at"] = now()
        extra_fields = {}

        if extra_updates:
            for k, v in extra_updates.items():
                if k in _TABLE_COLUMNS:
                    col_updates[k] = v
                else:
                    extra_fields[k] = v

        if extra_fields:
            col_updates["extra"] = merge_extra_json(conn, task_id, extra_fields)

        set_clause = ", ".join(f"{k} = ?" for k in col_updates)
        values = list(col_updates.values()) + [task_id]
        conn.execute(f"UPDATE tasks SET {set_clause} WHERE id = ?", values)

        # 记录日志 / Record transition log
        conn.execute(
            """
            INSERT INTO task_logs (task_id, from_status, to_status, trigger, note, created_at)
            VALUES (?, ?, ?, ?, ?, ?)
        """,
            (task_id, from_status, to_status, trigger, note, now()),
        )

        return from_status, to_status


def force_transition(task_id: str, to_status: str, note: str | None = None) -> str:
    """强制状态转换（绕过转换表验证，仅用于 watcher 恢复等紧急场景）
    Force state transition (bypass transition table validation, only for watcher recovery etc.).

    与直接 SQL 不同，此函数仍记录审计日志。
    Unlike direct SQL, this function still records audit logs.

    Args:
        task_id: 任务 ID / Task ID
        to_status: 目标状态 / Target state
        note: 备注 / Note

    Returns:
        from_status
    """
    import logging

    logging.getLogger(__name__).warning("force_transition：%s → %s（note=%s）", task_id, to_status, note)
    with atomic_transaction() as conn:
        row = conn.execute("SELECT status FROM tasks WHERE id = ?", (task_id,)).fetchone()
        if not row:
            raise ValueError(f"任务不存在：{task_id}")

        from_status = row["status"]
        conn.execute(
            "UPDATE tasks SET status = ?, updated_at = ?, started_at = ? WHERE id = ?",
            (to_status, now(), now(), task_id),
        )
        conn.execute(
            """
            INSERT INTO task_logs (task_id, from_status, to_status, trigger, note, created_at)
            VALUES (?, ?, ?, ?, ?, ?)
        """,
            (task_id, from_status, to_status, "force_transition", note, now()),
        )
        return from_status


def can_transition(task_id: str, trigger: str) -> bool:
    """检查是否可以执行某个 trigger
    Check if a given trigger can be executed."""
    from core.db import get_task

    task = get_task(task_id)
    if not task:
        return False

    workflow_name = task.get("workflow", "")
    try:
        from core import registry

        transitions = registry.build_transitions(workflow_name)
        if transitions:
            allowed = transitions.get(task["status"], [])
            return any(t == trigger for t, _ in allowed)
    except ImportError:
        pass
    except Exception as e:
        import logging

        logging.getLogger(__name__).warning("检查转换可行性失败：%s", e)

    return False


def get_available_triggers(task_id: str) -> list[str]:
    """获取当前状态下可用的 trigger 列表
    Get list of available triggers for the current state."""
    from core.db import get_task

    task = get_task(task_id)
    if not task:
        return []

    workflow_name = task.get("workflow", "")
    try:
        from core import registry

        transitions = registry.build_transitions(workflow_name)
        if transitions:
            return [t for t, _ in transitions.get(task["status"], [])]
    except ImportError:
        pass
    except Exception as e:
        import logging

        logging.getLogger(__name__).warning("获取可用触发器失败：%s", e)

    return []

File: src/core/state-machine.ts
import { getDb, getTask, now, TABLE_COLUMNS, PROTECTED_COLUMNS } from "./db";
import { emit } from "../daemon/event-bus";
import { appendTaskEvent } from "./task-logs";
import { appendTransition as appendManifestTransition } from "./manifest";

// ──────────────────────────────────────────────
// 类型定义
// ──────────────────────────────────────────────

/** 转换表：{ fromStatus: [[trigger, toStatus], ...] } */
export type TransitionTable = Record<string, [string, string][]>;

export class InvalidTransitionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InvalidTransitionError";
  }
}

export interface TransitionOptions {
  transitions: TransitionTable;
  note?: string;
  extraUpdates?: Record<string, unknown>;
}


// ──────────────────────────────────────────────
// 核心函数
// ──────────────────────────────────────────────

/**
 * 执行状态转换（原子性），返回 [fromStatus, toStatus]
 * @throws InvalidTransitionError 如果转换非法
 */
export function transition(
  taskId: string,
  trigger: string,
  opts: TransitionOptions
): [string, string] {
  const db = getDb();

  let fromStatus = "";
  let toStatus = "";

  db.transaction(() => {
    const task = db
      .query<{ status: string; extra: string | null }, [string]>(
        "SELECT status, extra FROM tasks WHERE id = ?"
      )
      .get(taskId);

    if (!task) {
      throw new InvalidTransitionError(`任务不存在：${taskId}`);
    }

    fromStatus = task.status;
    const available = opts.transitions[fromStatus] ?? [];
    const match = available.find(([t]) => t === trigger);

    if (!match) {
      throw new InvalidTransitionError(
        `非法转换：状态 "${fromStatus}" 不支持触发器 "${trigger}"`
      );
    }

    toStatus = match[1];
    const ts = now();

    // 分离列字段和 extra 字段
    const colUpdates: string[] = ["status = ?", "updated_at = ?"];
    const colValues: unknown[] = [toStatus, ts];

    // 如果目标状态以 "running" 开头，更新 started_at
    if (toStatus.startsWith("running")) {
      colUpdates.push("started_at = ?");
      colValues.push(ts);
    }

    const extraUpdatesForCol: Record<string, unknown> = {};
    const extraUpdatesForJson: Record<string, unknown> = {};

    if (opts.extraUpdates) {
      for (const [key, value] of Object.entries(opts.extraUpdates)) {
        if (key === "extra" || key === "status" || key === "updated_at" || PROTECTED_COLUMNS.has(key)) continue;
        if (TABLE_COLUMNS.has(key)) {
          extraUpdatesForCol[key] = value;
        } else {
          extraUpdatesForJson[key] = value;
        }
      }
    }

    // 追加列字段更新
    for (const [key, value] of Object.entries(extraUpdatesForCol)) {
      colUpdates.push(`${key} = ?`);
      colValues.push(value);
    }

    // 处理 extra JSON 合并
    if (Object.keys(extraUpdatesForJson).length > 0) {
      let currentExtra: Record<string, unknown> = {};
      try {
        currentExtra = task.extra ? JSON.parse(task.extra) : {};
      } catch {
        currentExtra = {};
      }
      const mergedExtra = { ...currentExtra, ...extraUpdatesForJson };
      colUpdates.push("extra = ?");
      colValues.push(JSON.stringify(mergedExtra));
    }

    colValues.push(taskId, fromStatus);
    const result = db.run(
      `UPDATE tasks SET ${colUpdates.join(", ")} WHERE id = ? AND status = ?`,
      colValues as Parameters<typeof db.run>[1]
    );

    if (result.changes === 0) {
      throw new InvalidTransitionError(
        `并发冲突：任务 "${taskId}" 状态已被其他进程修改`
      );
    }

    // 插入日志
    db.run(
      "INSERT INTO task_logs (task_id, from_status, to_status, trigger_name, note, created_at)" +
        " VALUES (?, ?, ?, ?, ?, ?)",
      [taskId, fromStatus, toStatus, trigger, opts.note ?? null, ts]
    );
  })();

  const transitionTs = now();
  appendManifestTransition(
    taskId,
    { from: fromStatus, to: toStatus, trigger, ts: transitionTs, note: opts.note ?? null },
    syncPatchFromTask(taskId),
  );

  emit({ type: "task:transition", payload: { taskId, from: fromStatus, to: toStatus, trigger } });
  appendTaskEvent(taskId, { type: "transition", from: fromStatus, to: toStatus, trigger, note: opts.note });

  return [fromStatus, toStatus];
}

/**
 * 从 DB 的最新任务状态派生 manifest 的同步 patch。
 * 用于 transition 完成后把 started_at、extra 等更新反映到 manifest。
 */
function syncPatchFromTask(taskId: string): {
  status: string;
  updated_at: string;
  started_at: string | null;
  extra: Record<string, unknown>;
  failure_count: number;
} | undefined {
  const t = getTask(taskId);
  if (!t) return undefined;
  const extra: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(t)) {
    if (TABLE_COLUMNS.has(k)) continue;
    extra[k] = v;
  }
  return {
    status: t.status,
    updated_at: t.updated_at,
    started_at: t.started_at,
    failure_count: t.failure_count,
    extra,
  };
}

/**
 * 强制状态转换（绕过转换表校验），保留审计日志。
 * 仅供 watcher 等系统组件在恢复卡死任务时使用。
 */
export function forceTransition(
  taskId: string,
  toStatus: string,
  note: string
): void {
  const db = getDb();

  let capturedFromStatus = "";
  let capturedTs = "";

  db.transaction(() => {
    const task = db
      .query<{ status: string }, [string]>(
        "SELECT status FROM tasks WHERE id = ?"
      )
      .get(taskId);

    if (!task) {
      throw new InvalidTransitionError(`任务不存在：${taskId}`);
    }

    capturedFromStatus = task.status;
    capturedTs = now();

    db.run(
      "UPDATE tasks SET status = ?, updated_at = ? WHERE id = ?",
      [toStatus, capturedTs, taskId]
    );

    db.run(
      "INSERT INTO task_logs (task_id, from_status, to_status, trigger_name, note, created_at)" +
        " VALUES (?, ?, ?, ?, ?, ?)",
      [taskId, capturedFromStatus, toStatus, "force_transition", note, capturedTs]
    );
  })();

  appendManifestTransition(
    taskId,
    { from: capturedFromStatus, to: toStatus, trigger: "force_transition", ts: capturedTs, note },
    syncPatchFromTask(taskId),
  );

  emit({ type: "task:transition", payload: { taskId, from: capturedFromStatus, to: toStatus, trigger: "force_transition" } });
  appendTaskEvent(taskId, { type: "transition", from: capturedFromStatus, to: toStatus, trigger: "force_transition", note });
}

/**
 * 检查是否可以从当前状态触发指定 trigger
 */
export function canTransition(
  taskId: string,
  trigger: string,
  opts: TransitionOptions
): boolean {
  const task = getTask(taskId);
  if (!task) return false;
  const available = opts.transitions[task.status] ?? [];
  return available.some(([t]) => t === trigger);
}

/**
 * 获取当前状态下所有可用的 trigger 列表
 */
export function getAvailableTriggers(
  taskId: string,
  opts: TransitionOptions
): string[] {
  const task = getTask(taskId);
  if (!task) return [];
  const available = opts.transitions[task.status] ?? [];
  return available.map(([t]) => t);
}
